File: src/comparator.ml
(*
MIT License

Copyright (c) 2018 by John Wickerson and Nathan Chong.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*)

(** Top-level entry for memory model comparison *)

open! Format
open! General_purpose

(** Check if we're on a Mac *)
let running_osx () =
  let ic = Unix.open_process_in "uname -s" in
  let os_name = input_line ic in
  ignore (Unix.close_process_in ic);
  if os_name = "Darwin" then true else false

(** Interface to the Alloy Java application *)
let run_alloy alloystar_dir xml_dir comparator_script iter solver quiet =
  let java_heap_size = opt "3g" iden (Sys.getenv_opt "JAVA_HEAP_SIZE") in
  let os =
    try Sys.getenv "OS" with Not_found ->
      if running_osx () then "x86-mac" else
        failwith
          "ERROR: Environment variable 'OS' not set. Try `source configure.sh`."
  in
  let legal_oses =
    ["x86-freebsd"; "x86-linux"; "x86-mac"; "x86-windows"; "amd64-linux"]
  in
  if not (List.mem os legal_oses) then
    failwith "Environment variable 'OS' must be set to one of [%s]."
      (String.concat "|" legal_oses);
  let solver_dir = MyFilename.concat [alloystar_dir; os] in
  let cmd =
    String.concat " "
      [sprintf "PATH=%s/%s:$PATH" alloystar_dir os;
       "timeout 2h";
       "java";
       "-Xmx" ^ java_heap_size;
       "-Djava.library.path=" ^ solver_dir;
       "-Dout=" ^ xml_dir;
       sprintf "-Dquiet=%B" quiet;
       "-Dsolver=" ^ solver;
       sprintf "-Dhigherorder=%b" (solver = "glucose");
       "-Dcmd=0";
       sprintf "-Diter=%B" iter;
       "-Dhash=../mk_hash";
       "edu/mit/csail/sdg/alloy4whole/RunAlloy";
       comparator_script]
  in
  Unix.chdir alloystar_dir;
  Sys.catch_break true;
  begin try
    let status = Unix.system cmd in
    match status with
    | Unix.WEXITED 0 -> () (* exited normally *)
    | Unix.WEXITED 124 -> printf "\nWARNING: Alloy timed out.\n"
    | Unix.WEXITED n -> failwith "Alloy failed with code %d." n
    | Unix.WSIGNALED n -> failwith "Alloy was killed with signal %d." n
    | Unix.WSTOPPED n -> failwith "Alloy was stopped with signal %d." n
  with Sys.Break ->
    printf "\nWARNING: Alloy was interrupted.\n"
  end;
  Sys.catch_break false

let customscript = ref None
let expect = ref (-1)
let batch = ref false
let default_solver = "glucose"
let solver = ref default_solver
let iter = ref false
let allowset = ref false
let filter_path = ref ""
             
let legal_solvers =
  ["sat4j"; "cryptominisat"; "plingeling"; "lingeling";
   "minisatprover"; "minisat"; "glucose"]
                 
let speclist =
  ["-customscript", Arg.String (set_option_ref customscript),
   "Custom comparator script (optional)";

   "-expect", Arg.Set_int expect,
   "Expect to find this many unique solutions (optional)";

   "-batch", Arg.Set batch,
   "Suppress GUI (optional)";

   "-solver", Arg.Set_string solver,
   sprintf "Which SAT solver to use (optional, default=%s). One of: [%s]"
     default_solver (String.concat "|" legal_solvers);

   "-iter", Arg.Set iter,
   "Find all solutions (optional)";

   "-allowset", Arg.Set allowset,
   "Generate the allow set too (optional)";

   "-filter", Arg.Set_string filter_path,
   "Only keep solutions that satisfy this model (optional)";
  ]

let get_args () =
  let usage_msg =
    "Top-level memalloy comparator.\nUsage: `comparator [options]`.\nOptions available:"
  in
  let bad_arg s =
    Arg.usage speclist usage_msg;
    failwith "Unexpected argument '%s'" s
  in
  let speclist = Global_options.speclist @ Pp_comparator.speclist @ speclist in
  Arg.parse speclist bad_arg usage_msg;
  if !iter then Pp_comparator.minimal := true

let check_args () =
  Pp_comparator.check_args ();
  if not (List.mem !solver (legal_solvers @ [""])) then
    failwith "The -solver flag must be set to one of [%s] (currently \"%s\")"
      (String.concat "|" legal_solvers) !solver

let append_line_to_file filename str =
  let oc = open_out_gen [Open_append; Open_creat] 0o755 filename in
  output_string oc (str ^ "\n");
  close_out oc

(** Interface to graphviz *)
let run_dot dot_file png_file =
  ignore (Sys.command (sprintf "dot -Tpng -o %s %s" png_file dot_file))

(** Put a handy "_latest" symlink in the memalloy/results directory *)
let create_latest_symlink results_dir =
  let top_results_dir = Filename.dirname results_dir in
  let latest_symlink =
    MyFilename.concat [top_results_dir; "_latest"]
  in
  if Sys.file_exists latest_symlink then
    Sys.remove latest_symlink;
  Unix.symlink results_dir latest_symlink

(** [copy_file src tgt] copies the file at [src] to the new path [tgt] *)
let copy_file src tgt =
  ignore (Sys.command (sprintf "cp %s %s" src tgt))
  
let main () =
  
  (* 1. Setup directory structure *)
  MyTime.start_timer ("setup");

  let ppc_config = Pp_comparator.build_config () in
  
  let memalloy_root = Filename.dirname Sys.executable_name in
  let top_results_dir = MyFilename.concat [memalloy_root; "results"] in
  if not (Sys.file_exists top_results_dir) then
    Unix.mkdir top_results_dir 0o755;
  
  let stamp = MyTime.timestamp () in  
  let results_dir = MyFilename.concat [top_results_dir; stamp] in
  let alloystar_dir = MyFilename.concat [memalloy_root; "alloystar"] in
  let xml_dir = MyFilename.concat [results_dir; "xml"] in
  let png_dir = MyFilename.concat [results_dir; "png"] in
  let dot_dir = MyFilename.concat [results_dir; "dot"] in
  let als_dir = MyFilename.concat [results_dir; "als"] in
  let lit_dir = MyFilename.concat [results_dir; "litmus"] in
  let c_dir = MyFilename.concat [results_dir; "C"] in

  List.iter (fun p -> Unix.mkdir p 0o755)
    [results_dir; xml_dir; png_dir; dot_dir; lit_dir; als_dir];

  if ppc_config.arch = Archs.C then
    Unix.mkdir c_dir 0o755;

  let allow_dir = MyFilename.concat [results_dir; "allow"] in
  let allow_xml_dir = MyFilename.concat [allow_dir; "xml"] in
  let allow_png_dir = MyFilename.concat [allow_dir; "png"] in
  let allow_dot_dir = MyFilename.concat [allow_dir; "dot"] in
  let allow_als_dir = MyFilename.concat [allow_dir; "als"] in
  let allow_lit_dir = MyFilename.concat [allow_dir; "litmus"] in

  if !allowset then
    List.iter (fun p -> Unix.mkdir p 0o755)
      [allow_dir; allow_xml_dir; allow_png_dir;
       allow_dot_dir; allow_lit_dir; allow_als_dir];

  create_latest_symlink results_dir;

  (* 2. Print a friendly description *)
  if ppc_config.description <> "" then
    printf "\"%s\"\n" ppc_config.description;

  (* 3. Generate comparator script *)
  let comparator_als = MyFilename.concat [results_dir; "comparator.als"] in  
  begin match !customscript with
  | Some s ->
     if Sys.file_exists s then copy_file s comparator_als
     else failwith "ERROR: Custom comparator script '%s' not found.\n" s
  | None ->
     let convert_to_als model =
       if Filename.check_suffix model ".cat" then
         ignore (Cat2als.als_of_file false model)
       else if Filename.check_suffix model ".als" then
         ()
       else
         failwith "ERROR: Unrecognised model type '%s'.\n" model
     in
     List.iter convert_to_als ppc_config.satisfies_paths;
     List.iter convert_to_als ppc_config.also_satisfies_paths;
     List.iter convert_to_als ppc_config.violates_paths;

     Pp_comparator.main ppc_config (Some comparator_als);
  end;
  MyTime.stop_timer();
  
  (* 4. Alloy solving *)
  MyTime.start_timer("alloy");
  run_alloy alloystar_dir xml_dir comparator_als !iter !solver false;
  let nsolutions = Array.length (Sys.readdir xml_dir) in
  MyTime.stop_timer();
  printf "Alloy found %d solutions.\n" nsolutions;

  if nsolutions = 0 then
    if !expect <= 0 then exit 0
    else failwith "ERROR: Expected %d solutions, found 0.\n" !expect;

  (* 5. Generate als output *)
  MyTime.start_timer("dump als");
  printf "Generating als files.\n%!";
  let generate_als dir xml_file =
    let test_name = Filename.chop_extension (Filename.basename xml_file) in
    let als_file = MyFilename.concat [dir; "als"; test_name ^ ".als"] in
    Gen.als_name := test_name;
    Gen.run xml_file als_file Gen.Als ppc_config.arch
  in
  MyFilename.iter (generate_als results_dir) xml_dir;
  MyTime.stop_timer();

  let filter model_to_satisfy als_dir xml_path =
    let test_name =
      Filename.chop_extension (Filename.basename xml_path)
    in
    let hint_file = MyFilename.concat [als_dir; test_name ^ ".als"] in
    let comparator_als =
      MyFilename.concat [results_dir; "tmp_comparator.als"]
    in
    let new_config = {
        Pp_comparator.default_config with
        satisfies_paths = [model_to_satisfy];
        hints = [hint_file];
        withinit = ppc_config.withinit;
        eventcount = ppc_config.eventcount;
        arch = ppc_config.arch;
      }
    in
    Pp_comparator.main new_config (Some comparator_als);
    let tmp_xml_file = MyFilename.concat [results_dir; "test_0.xml"] in
    if Sys.file_exists tmp_xml_file then
      Sys.remove tmp_xml_file;
    run_alloy alloystar_dir results_dir comparator_als false "sat4j" true;
    if Sys.file_exists tmp_xml_file then
      printf ".%!"
    else (
      printf "!%!";
      Sys.remove hint_file;
      Sys.remove xml_path
    )
  in

  (* 6. Filter executions to keep only those that satisfy the -filter model *)
  if !filter_path <> "" then (
    MyTime.start_timer("filtering");
    printf "Removing executions that violate the -filter model";
    MyFilename.iter (filter !filter_path als_dir) xml_dir;
    printf "\n";
    MyTime.stop_timer();
  );  
  
  (* 7. Generate allow-set by 'weakening' each forbidding execution. *)
  if !allowset then (
    MyTime.start_timer("allowset");
    Weaken.hashes_seen := [];
    Weaken.run allow_xml_dir xml_dir;
    MyTime.stop_timer();
    let nsolutions = Array.length (Sys.readdir allow_xml_dir) in
    printf "Generated %d allowed solutions.\n%!" nsolutions;
    MyFilename.iter (generate_als allow_dir) allow_xml_dir
  );

  (* 8. Validate that every execution in the allow-set is indeed allowed. *)
  if !allowset && List.length ppc_config.violates_paths = 1 then (
    MyTime.start_timer("check allow");
    printf "Checking validity of allow-tests";
    let violates_path = List.hd ppc_config.violates_paths in
    MyFilename.iter (filter violates_path allow_als_dir) allow_xml_dir;
    printf "\n";
    MyTime.stop_timer()
  );

  (* 9. Generate dot and png files from each execution. *)
  MyTime.start_timer("dump png");
  printf "Generating dot and png files.\n%!";
  let generate_dot_png dir xml_file =
    let test_name = Filename.chop_extension (Filename.basename xml_file) in
    let dot_file = MyFilename.concat [dir; "dot"; test_name ^ ".dot"] in
    Gen.run xml_file dot_file Gen.Dot ppc_config.arch;
    let png_file = MyFilename.concat [dir; "png"; test_name ^ ".png"] in
    run_dot dot_file png_file
  in
  MyFilename.iter (generate_dot_png results_dir) xml_dir;
  if !allowset then begin
      printf "Generating dot and png files for allow-set.\n%!";
      MyFilename.iter (generate_dot_png allow_dir) allow_xml_dir;
    end;
  MyTime.stop_timer();

  (* 10. Generate litmus files from each execution. *)
  MyTime.start_timer("dump litmus");
  printf "Generating litmus files.\n%!";
  let generate_lit dir xml_file =
    let all_file = MyFilename.concat [dir; "litmus"; "@all"] in
    let test_name = Filename.chop_extension (Filename.basename xml_file) in
    let lit_file = test_name ^ ".litmus" in
    let lit_path = MyFilename.concat [dir; "litmus"; lit_file] in
    append_line_to_file all_file lit_file;
    Gen.run xml_file lit_path Gen.Lit ppc_config.arch
  in
  MyFilename.iter (generate_lit results_dir) xml_dir;
  if !allowset then begin
      printf "Generating litmus files for allow-set.\n%!";
      MyFilename.iter (generate_lit allow_dir) allow_xml_dir
    end;
  MyTime.stop_timer();

  (* 11. Also build C witness if arch=C. *)
  if ppc_config.arch = Archs.C then begin
      MyTime.start_timer("dump C");
      printf "Generating C files.\n%!";
      let generate_c dir xml_file =
        let all_file = MyFilename.concat [dir; "C"; "@all"] in
        let test_name = Filename.chop_extension (Filename.basename xml_file) in
        let c_file = test_name ^ ".c" in
        let c_path = MyFilename.concat [dir; "C"; c_file] in
        append_line_to_file all_file c_file;
        Gen.run xml_file c_path Gen.C ppc_config.arch
      in
      MyFilename.iter (generate_c results_dir) xml_dir;
      if !allowset then begin
          printf "Generating C files for allow-set.\n%!";
          MyFilename.iter (generate_c allow_dir) allow_xml_dir
        end;
      MyTime.stop_timer();
    end;

  (* 12. Show outcome graphically. *)
  if not !batch then
    if nsolutions = 1 then begin
        let first_litmus_test = Array.get (Sys.readdir lit_dir) 0 in
        let first_litmus_path = MyFilename.concat [lit_dir; first_litmus_test] in
        ignore (Sys.command (sprintf "cat %s" first_litmus_path));
        if running_osx () then
          let first_png = Array.get (Sys.readdir png_dir) 0 in
          let first_png_path = MyFilename.concat [png_dir; first_png] in
          ignore (Sys.command (sprintf "open %s" first_png_path))
      end
    else
      if running_osx () then
        ignore (Sys.command (sprintf "open %s" png_dir));

  (* 13. Compare against expected number of solutions. *)
  if !expect >= 0 && !expect != nsolutions then
    failwith "ERROR: Expected %d solutions, found %d." !expect nsolutions;
  
  MyTime.report_timings ();

  exit 0

let _ =
  if MyStr.endswith Sys.argv.(0) "comparator" then begin
      get_args ();
      check_args ();
      main ()
    end

File: top/comparator.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# MIT License
# 
# Copyright (c) 2017 by Nathan Chong and John Wickerson
# 
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
# 
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

import argparse
import os
from shutil import copyfile, rmtree
import sys
from subprocess32 import call, Popen, TimeoutExpired
import platform
from timeit import default_timer as timer
import collections

import argparsing
from argparsing import TOOL_PATH
import setup_result_dir
import run_alloy
import remove_dups

def ext_of_file(f):
  _unused_basename, ext = os.path.splitext(f)
  return ext

def is_cat_file(f):
  return ext_of_file(f) == ".cat"

def is_als_file(f):
  return ext_of_file(f) == ".als"

def is_xml_file(f):
  return ext_of_file(f) == ".xml"

def main(argv=None):
  if argv is None:
    argv = sys.argv[1:]
  parser = argparse.ArgumentParser(description="Top-level memalloy comparator", conflict_handler="resolve")

  parser.add_argument("-verbose", action="store_true",
                      help="Option: show subcommands and use verbose output")
  parser.add_argument("-expect", type=int, default=None,
                      help="Expect to find this many unique solutions (optional)")
  parser.add_argument("-desc", type=str, help="Textual description (optional)")
  parser.add_argument("-batch", action="store_true",
                        help="Option: suppress GUI")
  parser.add_argument("-customscript", type=str, help="Custom comparator script (optional)")
  argparsing.add_setup_result_dir_args(parser)
  argparsing.add_gen_comparator_args(parser)
  argparsing.add_run_alloy_args(parser)
  argparsing.add_remove_dups_args(parser, is_standalone=False)
  args = parser.parse_args(argv)

  timing = collections.OrderedDict()

  # Stage 1: setup directory structure and generate comparator script
  start = timer()
  result_dir = setup_result_dir.main(args)
  xml_result_dir = os.path.join(result_dir, "xml")
  dot_result_dir = os.path.join(result_dir, "dot")
  png_result_dir = os.path.join(result_dir, "png")
  lit_result_dir = os.path.join(result_dir, "litmus")
  allow_result_dir = os.path.join(result_dir, "allow")
  allow_xml_result_dir = os.path.join(result_dir, "allow", "xml")
  allow_dot_result_dir = os.path.join(result_dir, "allow", "dot")
  allow_png_result_dir = os.path.join(result_dir, "allow", "png")
  allow_lit_result_dir = os.path.join(result_dir, "allow", "litmus")

  if args.desc:
    print "\"" + args.desc + "\""

  comparator_script = os.path.join(result_dir, "comparator.als")

  if args.customscript:

    if not os.path.exists(args.customscript):
      print "ERROR: Custom comparator script '%s' not found" % args.customscript
      return 1
    copyfile(args.customscript, comparator_script)

  else:

    # Deal with non-local models by copying them into models/_tmp_<n>.{cat,als}
    # and patching the command-line arguments
    nonlocal_model_map = {} # for temporaries
    all_models = args.satisfies + args.alsosatisfies + args.violates
    if args.filter: all_models = all_models + [args.filter]
    for model in all_models:
      model_path = os.path.abspath(model)
      if not os.path.dirname(model_path).startswith(argparsing.MEMALLOY_ROOT_DIR):
        tmp_cat = os.path.join("models", "_tmp_%d%s" % (len(nonlocal_model_map), ext_of_file(model)))
        nonlocal_model_map[model] = tmp_cat
        copyfile(model_path, os.path.join(argparsing.MEMALLOY_ROOT_DIR, tmp_cat))
        model = os.path.join(argparsing.MEMALLOY_ROOT_DIR, tmp_cat)
      if is_cat_file(model):
        if args.fencerels:
          cmd = [os.path.join(TOOL_PATH, "cat2als"), "-fencerels", model]
        else:
          cmd = [os.path.join(TOOL_PATH, "cat2als"), model]
        cmd.extend(["-u", str(args.unroll)])
        code = call(cmd)
        if code != 0:
          print "ERROR: Unable to convert cat file"
          return 1
      elif is_als_file(model):
        pass
      else:
        print "ERROR: Unrecognised model type [%s]" % model
        return 1

    if nonlocal_model_map:
      for k in ["satisfies", "alsosatisfies", "violates"]:
        for i,m in enumerate(args.__dict__[k]):
          if m in nonlocal_model_map:
            new_m = nonlocal_model_map[m]
            if args.verbose: print "Patching %s |-> %s" % (m, new_m)
            args.__dict__[k][i] = new_m

    cmd = [os.path.join(TOOL_PATH, "pp_comparator"), "-o", comparator_script]
    cmd.extend(argparsing.extract_gen_comparator_args(args))
    if args.verbose: print " ".join(cmd)
    code = call(cmd)
    if code != 0:
      print "ERROR: Generation of comparator script was unsuccessful"
      return 1
  end = timer()
  timing['setup'] = (end-start)
    
  #end if args.customscript

  # Stage 2: Alloy solving
  start = timer()
  args.comparator_script = comparator_script
  args.alloystar = "alloystar"
  args.xml_result_dir = xml_result_dir
  try:
    code = run_alloy.main(args)
  except KeyboardInterrupt:
    code = 0
    print "\nWARNING: Alloy was interrupted"
  except TimeoutExpired:
    code = 0
    print "\nWARNING: Alloy timeout (only partial solutions)"
  if code != 0:
    print "ERROR: Alloy was unsuccessful"
    return 1
  nsolutions = len([x for x in os.listdir(xml_result_dir) if is_xml_file(x)])
  end = timer()
  print "Alloy found %d solutions in %.2f sec" % (nsolutions, end-start)
  timing['alloy'] = (end-start)
  
  if nsolutions == 0:
    if args.expect and args.expect != 0:
      print "ERROR: Expected %d unique solutions, found 0" % args.expect
      return 1
    else:
      return 0
 
  # Stage 3: Remove duplicates
  print "Remove duplicates"
  start = timer()
  code = remove_dups.main(args)
  if code != 0:
    print "ERROR: Remove duplicates script was unsuccessful"
    return 1
  nsolutions = len([x for x in os.listdir(xml_result_dir) if x.endswith("_unique")])
  end = timer()
  print "Partitioned to %d unique solutions in %.2f sec" % (nsolutions, end-start)
  timing['rmdups'] = (end-start)

  # Stage 3a: Filter executions to remove those that are inconsistent according to the model given in the -filter parameter
  if args.filter:
    print "Filter executions"
    start = timer()
    hash_file = os.path.join(xml_result_dir, "hashes.txt")
    hash_file_keep = os.path.join(xml_result_dir, "hashes_keep.txt")
    hash_file_ignore = os.path.join(xml_result_dir, "hashes_ignore.txt")
    open(hash_file_keep, 'a').close()
    open(hash_file_ignore, 'a').close()
    nfiltered = 0    
    with open(hash_file) as f:
      for test_hash in f:
        test_hash = test_hash.strip()
        if not test_hash: continue
        xml_dir = os.path.join(xml_result_dir, "%s_unique" % test_hash)
        xml_files = [ x for x in os.listdir(xml_dir) if is_xml_file(x) ]
        assert 0 < len(xml_files)

        hint_file = os.path.join(result_dir, "test_%s.als" % test_hash)
        cmd = [os.path.join(TOOL_PATH, "gen"), "-Tals", \
                 "-name", "test_%s" % test_hash, \
                 "-o", hint_file, \
                 os.path.join(xml_dir, xml_files[0])]
        if args.verbose: print " ".join(cmd)
        code = call(cmd)
        if code != 0:
          print "ERROR: als generation was unsuccessful"
          return 1
        tmp_comparator_script = os.path.join(result_dir, "tmp_comparator.als")
        cmd = [os.path.join(TOOL_PATH, "pp_comparator"), \
                 "-arch", args.arch, \
                 "-satisfies", args.filter, \
                 "-events", str(args.events), \
                 "-hint", hint_file, \
                 "-o", tmp_comparator_script]
        if args.fencerels: cmd.extend(["-fencerels"])
        if args.verbose: print " ".join(cmd)
        code = call(cmd)
        if code != 0:
          print "ERROR: generating a comparator to check whether a generated execution should be filtered"
          return 1
        os.remove(hint_file)
        tmp_xml_result_dir = os.path.join(result_dir, "tmp_xml")
        if os.path.isdir(tmp_xml_result_dir):
          rmtree(tmp_xml_result_dir)
        os.mkdir(tmp_xml_result_dir)
        args.comparator_script = tmp_comparator_script
        args.xml_result_dir = tmp_xml_result_dir
        args.iter = False
        code = run_alloy.main(args)
        if code != 0:
          print "ERROR: checking whether a generated execution should be filtered was unsuccessful"
          return 1
        num_solutions = len([x for x in os.listdir(tmp_xml_result_dir) if is_xml_file(x)])
        if num_solutions == 1:
          print "Execution %s is consistent under baseline." % test_hash
          with open(hash_file_keep, "a") as f_keep:
            print >>f_keep, test_hash
        else:
          print "Execution %s is also inconsistent under baseline -- remove!" % test_hash
          with open(hash_file_ignore, "a") as f_ignore:
            print >>f_ignore, test_hash
          nfiltered = nfiltered + 1
    end = timer()
    print "Filtered down to %d solutions in %.2f sec" % (nsolutions - nfiltered, end-start)
    timing['filtering'] = (end-start)
  
  # Stage 3b: Generate allow-set
  if args.allowset:
    print "Generate allow-set"
    start = timer()
    cmd = [os.path.join(TOOL_PATH, "weaken"), \
             "-o", allow_xml_result_dir, \
             xml_result_dir]
    if args.verbose: print " ".join(cmd)
    code = call(cmd)
    if code != 0:
      print "ERROR: generating allowed executions was unsuccessful"
      return 1
    nsolutions = len([x for x in os.listdir(allow_xml_result_dir) if is_xml_file(x)])
    print "Constructed %d allowed variants" % nsolutions 
    print "Remove duplicates from allow-set"
    args.xml_result_dir = allow_xml_result_dir
    remove_dups.UNIQUE = {}
    code = remove_dups.main(args)
    if code != 0:
      print "ERROR: removing duplicates from allow-set was unsuccessful"
      return 1
    nsolutions = len([x for x in os.listdir(allow_xml_result_dir) if x.endswith("_unique")])
    end = timer()
    print "Partitioned to %d unique solutions in %.2f sec" % (nsolutions, end-start)
    timing['allowset'] = (end-start)
  
  # Stage 4: Generate litmus test output
  start = timer()
  result_dir_list = [result_dir]
  if args.allowset:
    result_dir_list = [result_dir, allow_result_dir]
  for my_result_dir in result_dir_list:
    print "Converting in %s" % my_result_dir
    hash_file = os.path.join(my_result_dir, "xml", "hashes.txt")
    #litmus_filenames = []
    with open(hash_file) as f:
      for test_hash in f:
        test_hash = test_hash.strip()
        if not test_hash: continue
        xml_dir = os.path.join(my_result_dir, "xml", "%s_unique" % test_hash)
        xml_files = [ x for x in os.listdir(xml_dir) if is_xml_file(x) ]
        assert 0 < len(xml_files)

        als_file = os.path.join(my_result_dir, "als", "test_%s.als" % test_hash)
        cmd = [os.path.join(TOOL_PATH, "gen"), "-Tals", \
                 "-name", "test_%s" % test_hash, \
                 "-o", als_file, \
                 os.path.join(xml_dir, xml_files[0])]
        if args.verbose: print " ".join(cmd)
        code = call(cmd)
        if code != 0:
          print "ERROR: als generation was unsuccessful"
          return 1
        
        dot_file = os.path.join(my_result_dir, "dot", "test_%s.dot" % test_hash)
        cmd = [os.path.join(TOOL_PATH, "gen"), "-Tdot", "-o", dot_file, os.path.join(xml_dir, xml_files[0])]
        if args.verbose: print " ".join(cmd)
        code = call(cmd)
        if code != 0:
          print "ERROR: dot generation was unsuccessful"
          return 1
      
        png_file = os.path.join(my_result_dir, "png", "test_%s.png" % test_hash)
        cmd = ["dot", "-Tpng", "-o", png_file, dot_file]
        if args.verbose: print " ".join(cmd)
        code = call(cmd)
        if code != 0:
          print "ERROR: png generation was unsuccessful"
          return 1

        litmus = "test_%s.litmus" % test_hash
        #litmus_filenames.append(litmus)

        archs = [args.arch]
        if args.arch == "HW":
          archs = ["ARM8", "PPC", "X86"]
        for arch in archs:
          lit_file = os.path.join(my_result_dir, "litmus", arch, litmus)
          cmd = [os.path.join(TOOL_PATH, "gen"), "-Tlit", \
                "-arch", arch, \
                "-o", lit_file, \
                os.path.join(xml_dir, xml_files[0])]
          if args.verbose: print " ".join(cmd)
          code = call(cmd)
          if code != 0:
            print "ERROR: litmus-test generation was unsuccessful"
            return 1

          
  # litmus7 @all
  archs = [args.arch]
  if args.arch == "HW":
    archs = ["ARM8", "PPC", "X86"]
  for arch in archs:
    
    litmus_filenames = []
    hashes_txt = "hashes_keep.txt" if args.filter else "hashes.txt"
    hash_file = os.path.join(result_dir, "xml", hashes_txt)
    with open(hash_file) as f:
      for test_hash in f:
        test_hash = test_hash.strip()
        litmus = "test_%s.litmus" % test_hash
        litmus_filenames.append(litmus)
    with open(os.path.join(result_dir, "litmus", arch, "@all"), "w+") as f:
      for test in litmus_filenames:
        print >>f, test
    if args.allowset:
      litmus_filenames = []
      hash_file = os.path.join(allow_result_dir, "xml", "hashes.txt")
      with open(hash_file) as f:
        for test_hash in f:
          test_hash = test_hash.strip()
          litmus = "test_%s.litmus" % test_hash
          litmus_filenames.append(litmus)
      with open(os.path.join(allow_result_dir, "litmus", arch, "@all"), "w+") as f:
        for test in litmus_filenames:
          print >>f, test

  # Stage 5: Check the allow-set
  if args.allowset and len(args.violates) == 1:
      print "Checking that the allow-set executions are allowed"
      allow_comparator_script = os.path.join(result_dir, "allow_comparator.als")
      hash_file = os.path.join(allow_result_dir, "xml", "hashes.txt")
      with open(hash_file) as f:
        for test_hash in f:
          sys.stdout.flush()
          test_hash = test_hash.strip()
          if not test_hash: continue
          hint_file = os.path.join(allow_result_dir, "als", "test_%s.als" % test_hash)
          cmd = [os.path.join(TOOL_PATH, "pp_comparator"), \
                "-arch", args.arch, \
                "-satisfies", (args.violates)[0], \
                "-events", str(args.events), \
                "-hint", hint_file, \
                "-o", allow_comparator_script]
          if args.fencerels: cmd.extend(["-fencerels"])
          if args.verbose: print " ".join(cmd)
          code = call(cmd)
          if code != 0:
            print "ERROR: generating a comparator to check that the allow-set executions are allowed was unsuccessful"
            return 1
          tmp_xml_result_dir = os.path.join(allow_result_dir, "tmp_xml")
          if os.path.isdir(tmp_xml_result_dir):
            rmtree(tmp_xml_result_dir)
          os.mkdir(tmp_xml_result_dir)
          args.comparator_script = allow_comparator_script
          args.xml_result_dir = tmp_xml_result_dir
          args.iter = False
          code = run_alloy.main(args)
          if code != 0:
            print "ERROR: checking that the allow-set executions are allowed was unsuccessful"
            return 1
          num_solutions = len([x for x in os.listdir(tmp_xml_result_dir) if is_xml_file(x)])
          if num_solutions == 1:
            print "Execution %s in allow-set is indeed consistent." % test_hash
          else:
            print "ERROR: execution %s in allow-set is inconsistent!" % test_hash
            return
          
  end = timer()
  timing['dump'] = (end-start)

  if platform.system() == "Darwin" and args.batch == False:
    if nsolutions == 1:
      f = os.path.join(lit_result_dir, args.arch, litmus_filenames[0])
      if os.path.isfile(f):
        Popen(["cat", os.path.join(lit_result_dir, f)])
      for f in os.listdir(png_result_dir):
        Popen(["open", os.path.join(png_result_dir, f)])
    else:
      Popen(["open", os.path.join(png_result_dir)])

  if args.expect:
    if args.expect != nsolutions:
      print "ERROR: Expected %d unique solutions, found %d" % (args.expect, nsolutions)
      return 1

  for (k,d) in timing.items():
    print "%s\t%.2f sec" % (k,d)
      
  return 0

if __name__ == '__main__':
  sys.exit(main())
